Sources/ClientRuntime/Interceptor/InterceptorContext.swift

```diff
@@ -0,0 +1,172 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/// The base type of all context objects passed to `Interceptor` methods.
+public protocol InterceptorContext: AnyObject {
+
+    /// The type of the transport message that will be transmitted by the operation being invoked.
+    associatedtype RequestType
+
+    /// The type of the transport message that will be received by the operation being invoked.
+    associatedtype ResponseType
+
+    /// The type of the attributes that will be available to all interceptors.
+    associatedtype AttributesType: HasAttributes
+
+    /// - Returns: The input for the operation being invoked.
+    func getInput() -> Any
+
+    /// - Returns: The attributes available in this interceptor context.
+    func getAttributes() -> AttributesType
+}
+
+public extension InterceptorContext {
```

Comment:
See my comment above, we should make the modeled input type an associated type of InterceptorContext so this conditional cast is unneeded